### PR TITLE
NEXT-23654 - Fix Product Listing returns falsy product sorting

### PR DESCRIPTION
**Pull request by LunaDotGit**
### 1. Why is this change necessary?
In the current state, the sorting of Products inside the Listing don't behave as expected. When products are sorted by a Field, it creates an `ORDER BY` Criteria e.g.
 ```SQL
ORDER BY MAX(`product`.`sales`) DESC
```
This example comes from the Sorting Criteria for Topseller.

The issue lies in the non-deterministic behaviour in combination with `LIMIT` and `OFFSET` as in the Pagination. In the example of Topseller products, if a product has a non unique value of `sales` it causes products to disappear in the listing on some pages or be duplicated (not on the same page, but comparing e.g. page 1 and 2), as it does not cut off the expected number of products provided by `OFFSET` in a deterministic way.

The issue can also be reproduced with other sorting criterias, if a value is not unique.

### 2. What does this change do, exactly?
The proposed solution, is to add a fallback `ORDER BY` Criteria (here `product`.`id` ASC) to re-introduce deterministic behaviour, as it causes the products to be sorted by `product`.`id` if the values provided in the Sorting Fields are not unique.

This is done by adding a "fallbackSortingField" to the `fields` in the `createDalSorting` function which is called within the ListingLoader. The priority is set to `-1` so it will be added to the end of the list.

The result in SQL is (in the example of the topseller criteria):
```SQL
ORDER BY MAX(`product`.`sales`) DESC, MIN(`product`.`id`)
```

### 3. Describe each step to reproduce the issue or behaviour.
1. Generate Demo Products on a Test-Environment (e.g. `bin/console framework:demodata`)
2. Sort by Topseller.
3. Open multiple Tabs and compare the pages, or trace the SQL Statement in the Symfony Toolbar (Doctrine -> # product-listing-route::loading::search-ids) E.g.:
```SQL
SELECT `product`.`id`, `product`.`product_number`, `product`.`auto_increment` FROM `product` LEFT JOIN `product` `product.parent` ON `product`.`parent_id` = `product.parent`.`id` AND `product`.`version_id` = `product.parent`.`version_id` LEFT JOIN `product_visibility` `product.visibilities` ON `product`.`visibilities` = `product.visibilities`.`product_id` AND `product`.version_id = `product.visibilities`.product_version_id LEFT JOIN `product_category_tree` `product.categoriesRo.mapping` ON `product`.`id` = `product.categoriesRo.mapping`.`product_id`  AND `product`.`version_id` = `product.categoriesRo.mapping`.`product_version_id` LEFT JOIN `category` `product.categoriesRo` ON `product.categoriesRo.mapping`.`category_id` = `product.categoriesRo`.`id`  AND `product.categoriesRo`.`version_id` = `product.categoriesRo.mapping`.`category_version_id` WHERE (`product`.`version_id` = 0x0FA91CE3E96A4BC2BE4BD9CE752C3425) AND ((((`product.visibilities`.`visibility` >= 30) AND `product.visibilities`.`sales_channel_id` = 0x270E711C315849ED8E34E450E48A664E AND IFNULL(`product`.`active`, `product.parent`.`active`) = 1) AND `product.categoriesRo`.`id` = 0x4484A74D380E4480AB448DF2EEBD6CF2 AND NOT (`product`.`display_group` <=> NULL))) GROUP BY `product`.`display_group` ORDER BY MAX(`product`.`sales`) DESC LIMIT 24 OFFSET 0;
```
4. If using SQL, adjust the Offset in 24 steps, and check for duplicate keys.

### 4. Please link to the relevant issues (if any).
https://issues.shopware.com/issues/NEXT-23654

### 5. Checklist

- [ ] I have rebased my changes to remove merge conflicts
- [ ] I have written tests and verified that they fail without my change
- [x] I have created a [changelog file](https://github.com/shopware/platform/blob/trunk/adr/2020-08-03-implement-new-changelog.md) with all necessary information about my changes
- [ ] I have written or adjusted the documentation according to my changes
- [x] This change has comments for package types, values, functions, and non-obvious lines of code
- [x] I have read the contribution requirements and fulfil them.

### CODE REVIEW

File: src/Core/Content/Product/SalesChannel/Sorting/ProductSortingEntity.php [hunk 1]
```diff
@@ -52,6 +52,7 @@ public function createDalSorting(): array
         $sorting = [];
 
         $fields = $this->fields;
+        array_push($fields, $this->getFallbackSortingField());
```
Comment (shyim) [inline src/Core/Content/Product/SalesChannel/Sorting/ProductSortingEntity.php]: can you add a small unit test for this?

Comment (shyim) [inline src/Core/Content/Product/SalesChannel/Sorting/ProductSortingEntity.php]:
```suggestion
        $fields[] = , $this->getFallbackSortingField();
```

**Comment by shyim** [comment]
Thank you for the PR. I added two small things :)

File: src/Core/Content/Product/SalesChannel/Sorting/ProductSortingEntity.php (hunk 1 above)
Comment (shyim) [reply src/Core/Content/Product/SalesChannel/Sorting/ProductSortingEntity.php]: I will talk with the team tomorrow and see if we have a better location to add the fallback sorting. We still have an issue in the backlog to add to all repository searches a fallback search as primary key.

Comment (LunaDotGit) [reply src/Core/Content/Product/SalesChannel/Sorting/ProductSortingEntity.php]: I accidentally commited that with the , still there :D

Comment (shyim) [reply src/Core/Content/Product/SalesChannel/Sorting/ProductSortingEntity.php]: Can you move the fallback sorting into the ProductListingFeaturesSubscriber so it's more obvious that we add a sorting :)

Inside the Struct, it feels like really hidden magic happening :)

https://github.com/shopware/platform/blob/d1ada242838f32dc5f9910aff7a63cf5b58053b2/src/Core/Content/Product/SalesChannel/Listing/ProductListingFeaturesSubscriber.php#L267

Comment (LunaDotGit) [reply src/Core/Content/Product/SalesChannel/Sorting/ProductSortingEntity.php]: I understand. I moved it there, hopefully that works for you :)

**Comment by OliverSkroblin** [comment]
Hey @TheAnimeGuru, sorry for keep you waiting and thanks for the report.

I fixed this in a previous version inside the ProductSortingEntity.php

https://github.com/shopware/shopware/blob/bb8f818d7a7c952e75ccacf633dc572c177bbba7/src/Core/Content/Product/SalesChannel/Sorting/ProductSortingEntity.php#L86

**Pull request closed without merging**